modernize service worker examples to use async/await (#13669)
mdn/content: files/en-us/web/api/service_worker_api/using_service_workers/index.md

You can see the [source code on GitHub](https://github.com/mdn/sw-test/), and [view the example live](https://mdn.github.io/sw-test/). The one bit we'll call out here is the promise (see [app.js lines 22-47](https://github.com/mdn/sw-test/blob/gh-pages/app.js#L22-L47)), which is a modified version of what you read about above, in the [Promises test demo](https://mdn.github.io/js-examples/promises-test/). It is different in the following ways:

> 

Author: Done!

    })
  );
});
```

We have opted for this fallback image because the only updates that are likely to fail are new images, as everything else is depended on for installation in the `install` event listener we saw earlier.

## Service Worker Navigation Preload

If enabled, the [navigation preload](/en-US/docs/Web/API/NavigationPreloadManager) feature starts downloading resources as soon as the fetch request is made, and in parallel with service worker bootup.
This ensures that download starts immediately on navigation to a page, rather than having to wait until the service worker has booted.
That delay happens relatively rarely, but is unavoidable when it does happen, and may be significant.

First the feature must be enabled during service worker activation, using {{domxref("NavigationPreloadManager.enable()", "registration.navigationPreload.enable()")}}:

```js
const enableNavigationPreload = async () => {
  if (self.registration.navigationPreload) {
    // Enable navigation preloads!
    await self.registration.navigationPreload.enable();
  }
};

self.addEventListener('activate', (event) => {
  event.waitUntil(enableNavigationPreload());
});
```

> Then use {{domxref("FetchEvent.preloadResponse", "event.preloadResponse")}} to wait for the preloaded resource to finish downloading in our `fetch` event handler.

Reviewer: ```suggestion
Then use {{domxref("FetchEvent.preloadResponse", "event.preloadResponse")}} to wait for the preloaded resource to finish downloading in the `fetch` event handler.
```

> Continuing the example from the previous sections, we insert the code to wait for the preloaded resource between the cache check and fetching from the network if that doesn't succeed.

Reviewer: ```suggestion
Continuing the example from the previous sections, we insert the code to wait for the preloaded resource after the cache check, and before fetching from the network if that doesn't succeed.
```

  }

  // Next try to get the resource from the network
  try {
    const responseFromNetwork = await fetch(request);
    // response may be used only once
    // we need to save clone to put one copy in cache
    // and serve second one
    putInCache(request, responseFromNetwork.clone());
    return responseFromNetwork;
  } catch (error) {
    const fallbackResponse = await caches.match(fallbackUrl);
    if (fallbackResponse) {
      return fallbackResponse;
    }
    // when even the fallback response is not available,
    // there is nothing we can do, but we must always
    // return a Response object
    return new Response('Network error happened', {
      status: 408,
      headers: { 'Content-Type': 'text/plain' },
    });
  }
};

self.addEventListener("fetch", (event) => {
  event.respondWith(
    cacheFirst({
      request: event.request,
      fallbackUrl: "/sw-test/gallery/myLittleVader.jpg",
    })
  );
});
```

We have opted for this fallback image because the only updates that are likely to fail are new images, as everything else is depended on for installation in the `install` event listener we saw earlier.

## Service Worker Navigation Preload

> If enabled, this feature starts downloading resources as soon as the fetch request is made, and in parallel with service worker bootup. This ensures that download starts immediately, rather than having to wait until the service worker has booted. That delay happens relatively rarely, but is unavoidable when it does happen, and significant.

Reviewer: Good to have more cross links to the related docs
```suggestion
If enabled, the [navigation preload](/en-US/docs/Web/API/NavigationPreloadManager) feature starts downloading resources as soon as the fetch request is made, and in parallel with service worker bootup.
This ensures that download starts immediately, rather than having to wait until the service worker has booted. That delay happens relatively rarely, but is unavoidable when it does happen, and may be significant.
```

> First the feature needs be enabled with `registration.navigationPreload.enable()`:

Reviewer: ```suggestion
First the feature needs be enabled with {{domxref("NavigationPreloadManager.enable(), registration.navigationPreload.enable()")}}:
```

```js
const enableNavigationPreload = async () => {
  if (self.registration.navigationPreload) {
    // Enable navigation preloads!
    await self.registration.navigationPreload.enable();
  }
};

self.addEventListener('activate', (event) => {
  event.waitUntil(enableNavigationPreload());
});
```

> Then we need to make use of `event.preloadResponse`, which is passed as `preloadResponsePromise` to the `cacheFirst` function. Instead of first doing a cache check and then fetching from the network if that doesn't succeed, there is a middle step. So the new process is:

Reviewer: ```suggestion
Then we need to make use of {{domxref("FetchEvent.preloadResponse", "event.preloadResponse")}}, which is passed as `preloadResponsePromise` to the `cacheFirst` function.
Instead of first doing a cache check and then fetching from the network if that doesn't succeed, there is a middle step.
So the new process is:
```

1. Check cache
2. Wait on `preloadResponsePromise`
3. If neither of these are defined then we go to the network.

```js
const addResourcesToCache = async (resources) => {
  const cache = await caches.open('v1');
  await cache.addAll(resources);
};

const putInCache = async (request, response) => {
  const cache = await caches.open('v1');
  await cache.put(request, response);
};

const cacheFirst = async ({ request, preloadResponsePromise, fallbackUrl }) => {
  // First try to get the resource from the cache
  const responseFromCache = await caches.match(request);
  if (responseFromCache) {
    return responseFromCache;
  }

>   // Next try to use the preloaded response, if it's there

Reviewer: ```suggestion
  // Next try to use (and cache) the preloaded response, if it's there
```

  const preloadResponse = await preloadResponsePromise;
  if (preloadResponse) {
    console.info('using preload response', preloadResponse);
    putInCache(request, preloadResponse.clone());
    return preloadResponse;
  }

  // Next try to get the resource from the network
  try {
    const responseFromNetwork = await fetch(request);
    // response may be used only once
    // we need to save clone to put one copy in cache
    // and serve second one
    putInCache(request, responseFromNetwork.clone());
    return responseFromNetwork;
  } catch (error) {
    const fallbackResponse = await caches.match(fallbackUrl);
    if (fallbackResponse) {
      return fallbackResponse;
    }
    // when even the fallback response is not available,
    // there is nothing we can do, but we must always
    // return a Response object
    return new Response('Network error happened', {
      status: 408,
      headers: { 'Content-Type': 'text/plain' },
    });
  }
};

const enableNavigationPreload = async () => {
  if (self.registration.navigationPreload) {
    // Enable navigation preloads!
    await self.registration.navigationPreload.enable();
  }
};

> 

Reviewer: ```suggestion
// Enable navigation preload in the 'activate' event
```

Fortunately, service workers' promise-based structure makes it trivial to provide further options towards success. We could do this:

> ```js

Reviewer: @SimonSiefke Would it be possible to update this and the sections below to reflect the existence of [NavigationPreloadManager](https://developer.mozilla.org/en-US/docs/Web/API/NavigationPreloadManager). I'm currently updating the docs for this in https://github.com/mdn/content/pull/13634 and the next step would be to update this doc and the example to use this feature. I can already see you are much better at JavaScript than me and this would be timely.

If enabled, this feature starts downloading resources as soon as the fetch request is made, and in parallel with service worker bootup. This ensures that download starts immediately, rather than having to wait until the service worker has booted. That delay happens relatively rarely, but is unavoidable when it does happen, and significant.

The change is that instead of first doing a cache check and then fetching from the network if that doesn't succeed, there is a middle step. So the new process is:
1. Check cache
2. Wait on [FetchEvent.preloadResponse](https://developer.mozilla.org/en-US/docs/Web/API/FetchEvent/preloadResponse)
3. If neither of these are true then you go to the network.

And of course special brownie points for caching the result of both the prefetch and the network request.

The  [FetchEvent.preloadResponse](https://developer.mozilla.org/en-US/docs/Web/API/FetchEvent/preloadResponse) is either a download or undefined, so you relatively easy to handle. The new PR shows how.

Reviewer: PS Note that this feature is experimental, but it is in Chrome with its massive market share, and in nightly for FF. I expect this will ship in a version or two.

Reviewer: And more, note that the response can vary in preload - it isn't necessarily the same as the normal response. Further, it could change based on the value of `setHeaderValue()`.
It doesn't "have to" though, so we could perhaps show the basic case where it does not and you just cache in the same way and then note that the value might change based on header too.

Author: Thanks for the feedback! Navigation preload would be a great addition for the Service Worker example: I have made another PR https://github.com/mdn/sw-test/pull/43 and also updated this PR.

Reviewer: Thanks! Example looks good. Once it has been merged (I don't have needed rights) we should update this section to include information about the preload response stuff too. I guess the timing depends a bit on when this gets merged. Is that just waiting on new review?

Reviewer 2: Example has been merged - possibly prematurely for the doc update to reflect - but that's a chicken egg situation we just have to deal with - feel free to go ahead and make the doc updates tho :)

## Service Worker Navigation Preload

If enabled, this feature starts downloading resources as soon as the fetch request is made, and in parallel with service worker bootup. This ensures that download starts immediately, rather than having to wait until the service worker has booted. That delay happens relatively rarely, but is unavoidable when it does happen, and significant.

First the feature needs be enabled with `registration.navigationPreload.enable()`:

```js
const enableNavigationPreload = async () => {
  if (self.registration.navigationPreload) {
    // Enable navigation preloads!
    await self.registration.navigationPreload.enable();
  }
};

self.addEventListener('activate', (event) => {
  event.waitUntil(enableNavigationPreload());
});
```

Then we need to make use of `event.preloadResponse`, which is passed as `preloadResponsePromise` to the `cacheFirst` function. Instead of first doing a cache check and then fetching from the network if that doesn't succeed, there is a middle step. So the new process is:

1. Check cache
2. Wait on `preloadResponsePromise`

> 3. If neither of these are defined then we go to the network.

Author: Here is the new part about Service Worker Navigation Preload. Please let me know what you think!
